### ISSUE TITLE
Check mode is ignored when using server side apply

### ISSUE THREAD

**Issue by flybyray** [ansible-collections/kubernetes.core#547]
##### SUMMARY
[Check mode is **not** supported](https://github.com/ansible-collections/kubernetes.core/commit/8182a3afd4e8bd87e59ea3a29b31735f3584961e#r91550401) in some cases. Those should be clarified.

For example [`server_side_apply` does not support check mode?](https://github.com/ansible-collections/kubernetes.core/pull/260)

##### ISSUE TYPE
- Documentation Report

##### COMPONENT NAME
k8s_module

##### ANSIBLE VERSION

```paste below
ansible [core 2.14.0]
  config file = /home/user/git-src/ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/user/git-src/venv/lib64/python3.9/site-packages/ansible
  ansible collection location = /home/user/git-src/collections/dev:/home/user/git-src/collections
  executable location = /home/user/git-src/venv/bin/ansible
  python version = 3.9.13 (main, Oct  3 2022, 14:07:57) [GCC 8.5.0 20210514 (Red Hat 8.5.0-15.0.1)] (/home/user/git-src/venv/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

**Comment by gravesm** [comment]
Reproducer:
```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - kubernetes.core.k8s:
        state: absent
        kind: Namespace
        name: testing

    - kubernetes.core.k8s:
        apply: true
        server_side_apply:
          field_manager: ansible
        definition:
          kind: Namespace
          apiVersion: v1
          metadata:
            name: testing
      check_mode: true

    - kubernetes.core.k8s_info:
        kind: Namespace
        name: testing
      register: result

    - assert:
        that: not result.resources
```

**Comment by flybyray** [comment]
Thanks for your playbook
I did the following to replay:

```shell
mkdir -p $HOME$(mktemp  -u -d)
cd $_
python3 --version
python3 -m venv venv
source venv/bin/activate
python3 -m pip install --upgrade pip
python3 -m pip -V
pip install ansible
pip install kubernetes
cat <<EOF > playbook.yml
#paste
ansible-playbook playbook.yml
```

And it failed:

```txt
user@host:~/tmp/tmp.yyKy2qGrkS$ mkdir -p $HOME$(mktemp  -u -d)
user@host:~/tmp/tmp.yyKy2qGrkS$ cd $_
user@host:~/tmp/tmp.jBlbqC24sl$ python3 --version
Python 3.9.13
user@host:~/tmp/tmp.jBlbqC24sl$ python3 -m venv venv
user@host:~/tmp/tmp.jBlbqC24sl$ source venv/bin/activate
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ python3 -m pip install --upgrade pip                                                                                                                                                                                             Collecting pip
  Using cached pip-22.3.1-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 20.2.4
    Uninstalling pip-20.2.4:
      Successfully uninstalled pip-20.2.4
Successfully installed pip-22.3.1
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ python3 -m pip -V
pip 22.3.1 from /home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/pip (python 3.9)
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ pip install ansible
Collecting ansible
  Using cached ansible-7.0.0-py3-none-any.whl (41.8 MB)
Collecting ansible-core~=2.14.0
  Using cached ansible_core-2.14.0-py3-none-any.whl (2.2 MB)
Collecting PyYAML>=5.1
  Using cached PyYAML-6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_12_x86_64.manylinux2010_x86_64.whl (661 kB)
Collecting packaging
  Using cached packaging-21.3-py3-none-any.whl (40 kB)
Collecting resolvelib<0.9.0,>=0.5.3
  Using cached resolvelib-0.8.1-py2.py3-none-any.whl (16 kB)
Collecting jinja2>=3.0.0
  Using cached Jinja2-3.1.2-py3-none-any.whl (133 kB)
Collecting cryptography
  Using cached cryptography-38.0.4-cp36-abi3-manylinux_2_28_x86_64.whl (4.2 MB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.1.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (25 kB)
Collecting cffi>=1.12
  Using cached cffi-1.15.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (441 kB)
Collecting pyparsing!=3.0.5,>=2.0.2
  Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)
Collecting pycparser
  Using cached pycparser-2.21-py2.py3-none-any.whl (118 kB)
Installing collected packages: resolvelib, PyYAML, pyparsing, pycparser, MarkupSafe, packaging, jinja2, cffi, cryptography, ansible-core, ansible
Successfully installed MarkupSafe-2.1.1 PyYAML-6.0 ansible-7.0.0 ansible-core-2.14.0 cffi-1.15.1 cryptography-38.0.4 jinja2-3.1.2 packaging-21.3 pycparser-2.21 pyparsing-3.0.9 resolvelib-0.8.1
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ pip install kubernetes
Collecting kubernetes
  Using cached kubernetes-25.3.0-py2.py3-none-any.whl (1.4 MB)
Collecting google-auth>=1.0.1                                                                                                                                                                                                                                                                        Using cached google_auth-2.14.1-py2.py3-none-any.whl (175 kB)
Collecting requests                                                                                                                                                                                                                                                                                  Using cached requests-2.28.1-py3-none-any.whl (62 kB)                                                                                                                                                                                                                                            Collecting certifi>=14.05.14                                                                                                                                                                                                                                                                         Using cached certifi-2022.9.24-py3-none-any.whl (161 kB)                                                                                                                                                                                                                                         Collecting requests-oauthlib                                                                                                                                                                                                                                                                         Using cached requests_oauthlib-1.3.1-py2.py3-none-any.whl (23 kB)                                                                                                                                                                                                                                Collecting websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0                                                                                                                                                                                                                                       Using cached websocket_client-1.4.2-py3-none-any.whl (55 kB)                                                                                                                                                                                                                                     Requirement already satisfied: setuptools>=21.0.0 in ./venv/lib/python3.9/site-packages (from kubernetes) (50.3.2)                                                                                                                                                                                 Requirement already satisfied: pyyaml>=5.4.1 in ./venv/lib/python3.9/site-packages (from kubernetes) (6.0)                                                                                                                                                                                         Collecting urllib3>=1.24.2                                                                                                                                                                                                                                                                           Using cached urllib3-1.26.13-py2.py3-none-any.whl (140 kB)                                                                                                                                                                                                                                       Collecting python-dateutil>=2.5.3                                                                                                                                                                                                                                                                    Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)                                                                                                                                                                                                                                 Collecting six>=1.9.0                                                                                                                                                                                                                                                                                Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting rsa<5,>=3.1.4
  Using cached rsa-4.9-py3-none-any.whl (34 kB)                                                                                                                                                                                                                                                    Collecting cachetools<6.0,>=2.0.0
  Using cached cachetools-5.2.0-py3-none-any.whl (9.3 kB)
Collecting pyasn1-modules>=0.2.1
  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)
Collecting idna<4,>=2.5
  Using cached idna-3.4-py3-none-any.whl (61 kB)
Collecting charset-normalizer<3,>=2
  Using cached charset_normalizer-2.1.1-py3-none-any.whl (39 kB)
Collecting oauthlib>=3.0.0
  Using cached oauthlib-3.2.2-py3-none-any.whl (151 kB)
Collecting pyasn1<0.5.0,>=0.4.6
  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)
Installing collected packages: pyasn1, websocket-client, urllib3, six, rsa, pyasn1-modules, oauthlib, idna, charset-normalizer, certifi, cachetools, requests, python-dateutil, google-auth, requests-oauthlib, kubernetes
Successfully installed cachetools-5.2.0 certifi-2022.9.24 charset-normalizer-2.1.1 google-auth-2.14.1 idna-3.4 kubernetes-25.3.0 oauthlib-3.2.2 pyasn1-0.4.8 pyasn1-modules-0.2.8 python-dateutil-2.8.2 requests-2.28.1 requests-oauthlib-1.3.1 rsa-4.9 six-1.16.0 urllib3-1.26.13 websocket-client-1.4.2
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ cat <<EOF > playbook.yml
#paste your playbook
EOF
(venv) user@host:~/tmp/tmp.jBlbqC24sl$ ansible-playbook playbook.yml
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **********************************************************************************************************************************************************************************************************************************************************************************

TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: Object of type bytes is not JSON serializable
fatal: [localhost]: FAILED! => {"changed": false, "module_stderr": "Traceback (most recent call last):\n  File \"/home/user/.ansible/tmp/ansible-tmp-1669830837.15446-2110118-243837735495182/AnsiballZ_k8s.py\", line 107, in <module>\n    _ansiballz_main()\n  File \"/home/user/.ansible/tmp/ansible-tmp-1669830837.15446-2110118-243837735495182/AnsiballZ_k8s.py\", line 99, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user/.ansible/tmp/ansible-tmp-1669830837.15446-2110118-243837735495182/AnsiballZ_k8s.py\", line 47, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.kubernetes.core.plugins.modules.k8s', init_globals=dict(_module_fqn='ansible_collections.kubernetes.core.plugins.modules.k8s', _modlib_path=modlib_path),\n  File \"/usr/lib64/python3.9/runpy.py\", line 225, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib64/python3.9/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib64/python3.9/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py\", line 486, in <module>\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py\", line 482, in main\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/modules/k8s.py\", line 459, in execute_module\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py\", line 807, in execute_module\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/common.py\", line 1114, in perform_action\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/k8sdynamicclient.py\", line 45, in apply\n  File \"/tmp/ansible_kubernetes.core.k8s_payload_buha78be/ansible_kubernetes.core.k8s_payload.zip/ansible_collections/kubernetes/core/plugins/module_utils/apply.py\", line 136, in k8s_apply\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/dynamic/client.py\", line 164, in server_side_apply\n    return self.request('patch', path, body=body, force_conflicts=force_conflicts, **kwargs)\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/dynamic/client.py\", line 55, in inner\n    resp = func(self, *args, **kwargs)\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/dynamic/client.py\", line 270, in request\n    api_response = self.client.call_api(\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/client/api_client.py\", line 348, in call_api\n    return self.__call_api(resource_path, method,\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/client/api_client.py\", line 180, in __call_api\n    response_data = self.request(\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/client/api_client.py\", line 407, in request\n    return self.rest_client.PATCH(url,\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/client/rest.py\", line 296, in PATCH\n    return self.request(\"PATCH\", url,\n  File \"/home/user/tmp/tmp.jBlbqC24sl/venv/lib64/python3.9/site-packages/kubernetes/client/rest.py\", line 168, in request\n    request_body = json.dumps(body)\n  File \"/usr/lib64/python3.9/json/__init__.py\", line 231, in dumps\n    return _default_encoder.encode(obj)\n  File \"/usr/lib64/python3.9/json/encoder.py\", line 199, in encode\n    chunks = self.iterencode(o, _one_shot=True)\n  File \"/usr/lib64/python3.9/json/encoder.py\", line 257, in iterencode\n    return _iterencode(o, 0)\n  File \"/usr/lib64/python3.9/json/encoder.py\", line 179, in default\n    raise TypeError(f'Object of type {o.__class__.__name__} '\nTypeError: Object of type bytes is not JSON serializable\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0

```

**Comment by flybyray** [comment]
Actually there is a [encode() function](https://github.com/ansible-collections/kubernetes.core/blob/main/plugins/module_utils/apply.py#L134) which is doing something completly wrong. The kubernetes python module never expects bytes as input.

When i apply this patch 
```diff
--- venv/lib/python3.9/site-packages/ansible_collections/kubernetes/core/plugins/module_utils/apply.py.bak      2022-11-30 19:06:47.776383773 +0100
+++ venv/lib/python3.9/site-packages/ansible_collections/kubernetes/core/plugins/module_utils/apply.py  2022-11-30 19:07:17.693755477 +0100
@@ -131,7 +131,7 @@
     existing, desired = apply_object(resource, definition)
     server_side = kwargs.get("server_side", False)
     if server_side:
-        body = json.dumps(definition).encode()
+        body = definition
         # server_side_apply is forces content_type to 'application/apply-patch+yaml'
         return resource.server_side_apply(
             body=body,
```

```shell
ANSIBLE_COLLECTIONS_SCAN_SYS_PATH=false ANSIBLE_COLLECTIONS_PATHS=$PWD/venv/lib/python3.9/site-packages ansible-playbook playbook.yml
```

result:
```txt
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **********************************************************************************************************************************************************************************************************************************************************************************
TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************changed: [localhost]

TASK [kubernetes.core.k8s_info] *******************************************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [assert] *************************************************************************************************************************************************************************************************************************************************************************************fatal: [localhost]: FAILED! => {
    "assertion": "not result.resources",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************************************************localhost                  : ok=3    changed=1    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by gravesm** [comment]
Thanks, it looks like there are actually two different problems here. I created a new issue for the other one: #548.

**Comment by flybyray** [comment]
Actually i am not quite sure if the kubernetes-client/python code is the reason for the new behavior.
[When I add `dry_run=kwargs.get("dry_run"),` here to the call to `resource.server_side_apply`.](https://github.com/ansible-collections/kubernetes.core/blob/main/plugins/module_utils/apply.py#L141)

Then check_mode works somehow. I just ask myself how this worked before.
For me it looks like kubernetes.core never used the contract of function calls to kubernetes-client/python correct.

**Comment by flybyray** [comment]
> Thanks, it looks like there are actually two different problems here. I created a new issue for the other one: #548.

Actually the above test playbook is broken for all supported kubernetes versions:
* `kubernetes.core==2.3.2` with `kubernetes in [ 24.2.0, 23.6.0, 23.3.0, 22.6.0, 21.7.0, 20.13.0, 19.15.0 ]`. The failed task message just changed.
  ```txt
  TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************
  fatal: [localhost]: FAILED! => {"changed": false, "error": 0, "msg": "Namespace testing: Failed to apply object: None", "reason": "Cannot prepare a request message for provided\n                             arguments. Please check that your arguments match\n                             declared content type.", "status": 0}
  ```

**Comment by flybyray** [comment]
My current workarround to make the above playbook work is by installing a patched version of kubernetes.core.

```yaml
collections:
  - source: https://github.com/ndgit/kubernetes.core,2.3.2-flybyray-patch-1
    type: git
```

```shell
ANSIBLE_COLLECTIONS_SCAN_SYS_PATH=false ANSIBLE_COLLECTIONS_PATHS=$PWD/venv/lib/python3.9/site-packages ansible-galaxy collection install -r <(echo $'collections:\n  - source: https://github.com/ndgit/kubernetes.core,2.3.2-flybyray-patch-1-1\n    type: git')

ANSIBLE_COLLECTIONS_SCAN_SYS_PATH=false ANSIBLE_COLLECTIONS_PATHS=$PWD/venv/lib/python3.9/site-packages ansible-playbook playbook.yml
```

result:
```txt
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [localhost] **********************************************************************************************************************************************************************************************************************************************************************************
TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [kubernetes.core.k8s] ************************************************************************************************************************************************************************************************************************************************************************changed: [localhost]

TASK [kubernetes.core.k8s_info] *******************************************************************************************************************************************************************************************************************************************************************ok: [localhost]

TASK [assert] *************************************************************************************************************************************************************************************************************************************************************************************ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

PLAY RECAP ****************************************************************************************************************************************************************************************************************************************************************************************localhost                  : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

**Comment by gravesm** [comment]
This was fixed by #549.
